### PR TITLE
Add initial test script

### PR DESCRIPTION
**Pull request by alexguo8**
## Notion ticket link

[Ticket Name](https://www.notion.so/uwblueprintexecs/Task-Board-db95cd7b93f245f78ee85e3a8a6a316d)

## Implementation description

- Currently tests are sequential - we discussed pre-loading data earlier, but I opted not to do that for the first iteration since we need to pre-load firebase data as well, need to figure out a way to pre-load data that doesn’t just mimic the starter code implementation 
- It does tell you where the test fails, but it will just skip the rest of the tests for that suite
- So these are more like end-to-end tests for each group of endpoints
- One benefit is that they are self-cleaning (it will create data in the database and delete all of it when the test finishes)

## Steps to test

1. Add EMAIL=<login_email> and PASSWORD=<password> to root .env file (your starter code login credentials)
2. Run `docker-compose up --build` in the starter-code-v2 repo
3. Run pytest in the tests folder of this repo, with command line arguments according to the stack combination
--lang python (if running Python)
--api graphql (if running GraphQL)
--auth true (if using auth)
--fs true (if using file storage)
4. Verify that all test suites pass, and that no new data is left in the database

## What should reviewers focus on?

- Is there any part that can be written in a cleaner way?
- Not sure how to do logout for a new user since they need to verify their email first, should we instead logout the admin user we run the script with? (requires either enforcing an order so logout comes last or performing login before each test suite)

## Checklist

- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR
- [ ] I have updated the version number in `package.json` according to Semantic Versioning specs ([semver](https://semver.org)) if I will be publishing these changes to the npm registry

### CODE REVIEW

File: tests/conftest.py
Comment (sherryhli) [inline tests/conftest.py]: I think we can specify the valid choices

```suggestion
    parser.addoption("--lang", action="store", default="ts", choices=["ts", "python"])
```

File: tests/conftest.py
Comment (sherryhli) [inline tests/conftest.py]: Nit: could simplify so that we don't need to explicitly pass a value of "true". It's true if the flag is supplied, otherwise false

```suggestion
    parser.addoption("--auth", action="store_true")
    parser.addoption("--fs", action="store_true")
```

File: tests/test_auth.py
Comment (sherryhli) [inline tests/test_auth.py]: The register response should contain `id`, `role`, and `accessToken` fields as well. We can check that `role == "User"` and that the `id` and `accessToken` are present (we won't know their exact values)

Same thing for the graphql test :)

File: tests/test_auth.py
Comment (sherryhli) [inline tests/test_auth.py]: To avoid duplicating test data everywhere, we could use the inflection library to convert cases
https://inflection.readthedocs.io/en/latest/#inflection.camelize

File: tests/test_entity.py
Comment (sherryhli) [inline tests/test_entity.py]: Maybe check that `fileName` is present in the response as well?

File: tests/test_entity.py
Comment (sherryhli) [inline tests/test_entity.py]: Maybe a little redundant, but could do an assert on the `get_entities` response body

File: tests/test_entity_gql.py
Comment (sherryhli) [inline tests/test_entity_gql.py:115]: Nice 🔥

File: tests/test_user_gql.py
Comment (sherryhli) [inline tests/test_user_gql.py]: Let's make the backend base URL configurable by setting it as an env variable (starter code users can run these tests as well and they may choose to change the port)

File: tests/conftest.py
Comment (sherryhli) [inline tests/conftest.py]: > Not sure how to do logout for a new user since they need to verify their email first, should we instead logout the admin user we run the script with? (requires either enforcing an order so logout comes last or performing login before each test suite)

Let's try logging out the admin user. If we scope the `auth_header` fixture to `module` instead of `session` I think we can ensure that all tests will receive a valid access token regardless of when the logout test is executed

File: tests/test_entity.py
Comment (sherryhli) [inline tests/test_entity.py:81]: Note: release branch hasn't been updated with your changes in [#133](https://github.com/uwblueprint/starter-code-v2/pull/133) yet so this will fail when using TypeScript backend (which still returns 204)

**Review by sherryhli — APPROVED** [review]
Awesome work! Thanks so much for taking the lead on this effort 😄 All comments are really on nice-to-haves or nits, up to you if you want to address them

Did not exhaustively test the combinations, will have the opportunity to do that when we release the recent patches to starter code


**General comment:**
What do you think about moving tests to the starter-code-v2 repo instead? We talked about making these available to users to run as a sanity check after setup and it would be more accessible there since it'd be cloned with the rest of the code

**Comment by sherryhli** [comment]
I also seem to be getting 2 "Verify your email" emails sent to infra@ per test run, which suggests that the register user endpoint is being hit twice...though I don't see that happening in the code. Any idea why? 🤔

I used this combination:
```
 python -m pytest --auth true --fs true
```

File: tests/conftest.py
Comment (sherryhli) [inline tests/conftest.py]: We'll need to run the login mutation instead if using GraphQL

**Review by sherryhli — CHANGES_REQUESTED** [review]
Whoops, found a small issue with the auth_header fixture

**Comment by alexguo8** [comment]
> I also seem to be getting 2 "Verify your email" emails sent to infra@ per test run, which suggests that the register user endpoint is being hit twice...though I don't see that happening in the code. Any idea why? 🤔
> 
> I used this combination:
> 
> ```
>  python -m pytest --auth true --fs true
> ```

The create user endpoint also sends a verification email

**Comment by alexguo8** [comment]
> **General comment:** What do you think about moving tests to the starter-code-v2 repo instead? We talked about making these available to users to run as a sanity check after setup and it would be more accessible there since it'd be cloned with the rest of the code

I think that's a good idea, these tests do require some setup steps first though that we should document (eg. must insert an admin user into the DB to test with and set the env values, firebase console should be set up). Also, should we leave the test user email as the infra email?

Should I hold off on merging this and copy the changes over to starter-code-v2 or should I merge this for now and move it later?

**Comment by alexguo8** [comment]
Changes addressed, update env values to TEST_SCRIPT_EMAIL, TEST_SCRIPT_PASSWORD, and TEST_SCRIPT_BACKEND_URL (this is so it's more clear if we use the script in starter-code-v2)

Apologies for the large diff, most of it should be from updating the backend URL

**Review by sherryhli — DISMISSED** [review]
Awesome, LGTM! 🚢

> I think that's a good idea, these tests do require some setup steps first though that we should document (eg. must insert an admin user into the DB to test with and set the env values, firebase console should be set up). Also, should we leave the test user email as the infra email?
> 
> Should I hold off on merging this and copy the changes over to starter-code-v2 or should I merge this for now and move it later?

I think we can move over everything to starter-code-v2 right now. Most of the set up steps (create admin user, create Firebase project) are actually part of the starter code "Getting Started" guide already. The only new steps would be to add those `TEST_SCRIPT_` .env values which I can add to the guide :)

Good point about the infra@ email, maybe we should make that configurable via a env variable as well, e.g. `TEST_SCRIPT_NEW_USER_EMAIL`?

**Review by sherryhli — APPROVED** [review]
Looks good! Let's close this after merging https://github.com/uwblueprint/starter-code-v2/pull/137 though

**Comment by alexguo8** [comment]
Moved to main starter-code-v2 repo https://github.com/uwblueprint/starter-code-v2/pull/137

**Pull request closed without merging**
